### PR TITLE
Nit-added-tokens

### LINKED ISSUE #26500: Tokenizer pad token not saved with save_pretrained
### System Info

works on 4.33.3 (with tokenizers==0.13.3), fails on main (with tokenizers==0.14.0)

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```
import transformers
tokenizer = transformers.AutoTokenizer.from_pretrained('stabilityai/FreeWilly2')
tokenizer.pad_token_id = 0
print(tokenizer.pad_token)
tokenizer.save_pretrained('/tmp/tok_test')
tok = transformers.AutoTokenizer.from_pretrained('/tmp/tok_test')
print(tok.pad_token)
```

### Expected behavior

expected:

```
>>> print(tok.pad_token)
<unk>
```

actual:

```
>>> print(tok.pad_token)
Using pad_token, but it is not set yet.
None
```

### LINKED ISSUE #26536: Tokenizer AddedToken load from file bug
### System Info

- `transformers` version: 4.34.0.dev0
- Platform: macOS-13.6-arm64-arm-64bit
- Python version: 3.9.6
- Huggingface_hub version: 0.16.4
- Safetensors version: 0.3.2
- Accelerate version: 0.23.0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.0.1 (False)

### Who can help?

@ArthurZucker and @younesbelkada

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

```python
import tokenizers

from transformers import (
    AutoTokenizer
)

START_TOKEN = "<|im_start|>"
END_TOKEN = "<|im_end|>"
tokenizer_path = 'model/Llama-2-7b-hf'
complete_tokenizer_path = 'model/complete_tokenizer'
use_fast = False
legacy = True

tokenizer = AutoTokenizer.from_pretrained(
    tokenizer_path,
    use_fast=use_fast,
    legacy=legacy,
)
tokenizer.add_tokens([
    tokenizers.AddedToken(START_TOKEN, lstrip=False, rstrip=False, normalized=False, special=True),
    tokenizers.AddedToken(END_TOKEN, lstrip=False, rstrip=False, normalized=False, special=True),
])

before = tokenizer.encode(f'{START_TOKEN}\n')

tokenizer.save_pretrained('model/complete_tokenizer')

tokenizer = AutoTokenizer.from_pretrained(
    complete_tokenizer_path,
    use_fast=use_fast,
    legacy=legacy,
    model_max_length=2048,
)

after = tokenizer.encode(f'{START_TOKEN}\n')
assert before == after, (before, after)
```

### Expected behavior

The saved tokenizer files looks good but after `from_pretrained`, the customized AddedToken settings for rstrp and lstrip set to True are not being recognized as expected.

The saved tokenizer_config file:
``` json
{
  "add_bos_token": true,
  "add_eos_token": false,
  "added_tokens_decoder": {
    "0": {
      "content": "<unk>",
      "lstrip": false,
      "normalized": true,
      "rstrip": false,
      "single_word": false,
      "special": true
    },
    "1": {
      "content": "<s>",
      "lstrip": false,
      "normalized": true,
      "rstrip": false,
      "single_word": false,
      "special": true
    },
    "2": {
      "content": "</s>",
      "lstrip": false,
      "normalized": true,
      "rstrip": false,
      "single_word": false,
      "special": true
    },
    "32000": {
      "content": "<|im_start|>",
      "lstrip": false,
      "normalized": false,
      "rstrip": false,
      "single_word": false,
      "special": true
    },
    "32001": {
      "content": "<|im_end|>",
      "lstrip": false,
      "normalized": false,
      "rstrip": false,
      "single_word": false,
      "special": true
    }
  },
  "additional_special_tokens": [
    "<|im_start|>",
    "<|im_end|>"
  ],
  "bos_token": "<s>",
  "clean_up_tokenization_spaces": false,
  "eos_token": "</s>",
  "legacy": true,
  "model_max_length": 1600,
  "pad_token": null,
  "padding_side": "right",
  "sp_model_kwargs": {},
  "spaces_between_special_tokens": false,
  "tokenizer_class": "LlamaTokenizer",
  "tokenizer_file": null,
  "unk_token": "<unk>",
  "use_default_system_prompt": true
}


```

print(tokenizer) will show :
```
LlamaTokenizer(name_or_path='model/complete_tokenizer', vocab_size=32000, model_max_length=2048, is_fast=False, padding_side='right', truncation_side='right', special_tokens={'bos_token': '<s>', 'eos_token': '</s>', 'unk_token': '<unk>', 'additional_special_tokens': ['<|im_start|>', '<|im_end|>']}, clean_up_tokenization_spaces=False),  added_tokens_decoder={
        0: AddedToken("<unk>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
        1: AddedToken("<s>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
        2: AddedToken("</s>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
        32000: AddedToken("<|im_start|>", rstrip=True, lstrip=True, single_word=False, normalized=False, special=True),
        32001: AddedToken("<|im_end|>", rstrip=True, lstrip=True, single_word=False, normalized=False, special=True),
}
```
This only effect slow tokenizer

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Fixes #26500, fixes #26536

### CODE REVIEW

File: src/transformers/tokenization_utils_base.py [hunk 1]
```diff
@@ -2382,8 +2387,8 @@ def save_pretrained(
         tokenizer_config = copy.deepcopy(self.init_kwargs)
 
         # TODO: Ensure the modified attributes (those are also in the __init__ kwargs) will give identical tokenizers
-        # target_keys = self.init_kwargs.keys()
-        target_keys = ["model_max_length", "clean_up_tokenization_spaces", "additional_special_tokens"]
+        target_keys = list(self.init_kwargs.keys())
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:2390]: when saving; we should overwrite the init_kwargs with the content of self. Don't know why it was not the case before

File: src/transformers/tokenization_utils_base.py [hunk 2]
```diff
@@ -2227,7 +2232,7 @@ def _from_pretrained(
             if added_tokens_file is not None:
                 with open(added_tokens_file, encoding="utf-8") as added_tokens_handle:
                     added_tok_encoder = json.load(added_tokens_handle)
-                # legacy: we have to init with (rstrip=True, lstrip=True)
+                # legacy: we have to init with (rstrip=True, lstrip=True) (if the token is new? Failing test)
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:2235]: Might have to update this. The tests are shitty and the default is biting us

File: src/transformers/tokenization_utils_base.py [hunk 3]
```diff
@@ -2199,6 +2199,11 @@ def _from_pretrained(
 
                 if isinstance(token, AddedToken):
                     added_tokens_decoder[int(idx)] = token
+                    if str(token) in additional_special_tokens:
+                        # at this point if the token is in `additional_special_tokens` as an str, should be updated
+                        additional_special_tokens.remove(str(token))
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:2204]: Only use the default legacy values for AddedToken if the token is not already in the added tokens decoder

**Review by LysandreJik — APPROVED** [review]
LGTM

**Review by LysandreJik — APPROVED** [review]

**Comment by ArthurZucker** [comment]
A small benchmark on the `get_added_vocab()`:
```python 
from transformers import AutoTokenizer
import time 
tokenizer = AutoTokenizer.from_pretrained("facebook/nllb-moe-54b")
start = time.time();tokenizer.get_added_vocab();print(time.time()-start)
>>> 0.17021536827087402

start = time.time();{k.content: v for v, k in sorted(tokenizer.added_tokens_decoder.items(), key=lambda item: item[0])};print(time.time()-start)
>>>0.0054759979248046875
```

File: src/transformers/tokenization_utils_base.py [hunk 4]
```diff
@@ -2196,9 +2199,13 @@ def _from_pretrained(
...
                 if isinstance(token, dict):
                     token = AddedToken(**token)
-
                 if isinstance(token, AddedToken):
                     added_tokens_decoder[int(idx)] = token
+                    if str(token) in additional_special_tokens:
+                        # at this point the token is in `additional_special_tokens` as an str, let's add the AddedToken info
+                        additional_special_tokens.remove(str(token))
+                    if token.special and token not in additional_special_tokens:
+                        additional_special_tokens.append(token)
```
Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:2208]: This might be the culprit for MAIN failing

Comment (ArthurZucker) [inline src/transformers/tokenization_utils_base.py:2208]: cc @LysandreJik here

**Pull request merged**